File: 05_Graph_Search/Level3/2644.cpp
// Copyright@2023 Jihoon Lucas Kim <jihoon.lucas.kim@gmail.com>
// 촌수계산
// https://www.acmicpc.net/problem/2644
// 힌트
// 1. 부모-자식, 자식-부모 와같이 특별히 순서는 상관 없다.
// 2. DFS를 통해 목표로 하는 노드를 찾는데 걸리는 스텝이 촌수계산의 답이 됩니다.
#include <iostream>
#include <algorithm>
#include <vector>
#include <queue>
using namespace std;

int n, m, x, y;
int a, b;

vector<vector<bool> > table;

int BFS(int a)
{
    vector<bool> check(n + 1, true);
    queue<int> q;
    q.push(a);
    int answer = 0;
    while (q.size())
    {
        int size = q.size();
        for (int i = 0; i < size; i++)
        {
            int curr = q.front();
            q.pop();
            if (curr == b)
                return answer;
            if (check[curr])
            {
                check[curr] = false;
                for (int j = 1; j < n + 1; j++)
                {
                    if (table[curr][j] && check[j])
                    {
                        q.push(j);
                    }
                }
            }
        }
        answer += 1;
    }
    return -1;
}

int main()
{
    cin >> n;
    cin >> a >> b;
    cin >> m;
    table = vector<vector<bool> >(n + 1, vector<bool>(n + 1, false));
    for (int i = 0; i < m; i++)
    {
        cin >> x >> y;
        table[x][y] = true;
        table[y][x] = true;
    }
    int answer = BFS(a);
    cout << answer << endl;
    return 0;
}

File: 05_Graph_Search/Level3/2644.py
# Copyright@2023 Jihoon Lucas Kim <jihoon.lucas.kim@gmail.com>
# 촌수계산
# https://www.acmicpc.net/problem/2644
# 힌트
# 1. 부모-자식, 자식-부모 와같이 특별히 순서는 상관 없다.
# 2. BFS를 통해 목표로 하는 노드를 찾는데 걸리는 스텝이 촌수계산의 답이 됩니다.
import sys
from collections import deque


def bfs(s, e):
    ans = 0

    q = deque()
    q.append(s)
    check[s] = True

    while len(q) != 0:
        q_size = len(q)
        for i in range(q_size):
            current = q.popleft()

            if current == end:
                return ans

            for j in range(1, N + 1):
                if not check[j] and table[current][j]:
                    check[j] = True
                    q.append(j)

        ans += 1
    return -1


if __name__ == "__main__":
    N = int(sys.stdin.readline())

    start, end = map(int, sys.stdin.readline().split())
    M = int(sys.stdin.readline())

    check = [False] * (N + 1)
    table = [[False] * (N + 1) for _ in range(N + 1)]

    for i in range(M):
        x, y = map(int, sys.stdin.readline().split())
        table[x][y] = True
        table[y][x] = True

    answer = bfs(start, end)

    print(answer)